src/composeappengine.cc

```diff
@@ -42,46 +43,42 @@ void ComposeAppEngine::remove(const App& app) {
 }
 
 bool ComposeAppEngine::isRunning(const App& app) const {
-  bool cmd_res{false};
-  std::string cmd_output;
-  std::tie(cmd_res, cmd_output) = cmd("cat " + (appRoot(app) / ComposeFile).string());
-  if (!cmd_res) {
-    LOG_WARNING << "Failed to parse App config: " << app.name;
-    return false;
-  }
-
-  // calculate a number of container images that a given App consists of
-  std::string::size_type find_pos{0};
-  std::string::size_type line_pos{0};
-  int expected_container_number{0};
+  try {
+    ComposeInfo compose((appRoot(app) / ComposeFile).string());
+    std::vector<Json::Value> services;
 
-  while ((find_pos = cmd_output.find("image:", line_pos)) != std::string::npos) {
-    if (cmd_output.substr(line_pos, (find_pos - line_pos)).find_first_of('#') == std::string::npos) {
-      ++expected_container_number;
+    if (!compose.getServices(services)) {
+      LOG_WARNING << "App: " << app.name << ", no services!";
+      return false;
     }
-    line_pos = cmd_output.find('\n', find_pos);
-    ++find_pos;
-  }
-
-  // Get a number of running container images
-  std::tie(cmd_res, cmd_output) =
-      cmd(docker_ + "ps -q --filter=status=running --filter=label=com.docker.compose.project=" + app.name);
-  if (!cmd_res) {
-    LOG_WARNING << "Failed to get a list of App's containers: " << app.name;
-    return false;
-  }
 
-  int running_container_number =
-      std::count_if(cmd_output.begin(), cmd_output.end(), [](const char& symbol) { return symbol == '\n'; });
-
-  if (running_container_number < expected_container_number) {
-    LOG_DEBUG << "Number of running containers is less than a number of images specified in the compose file"
-              << "; App: " << app.name << "; expected container number: " << expected_container_number
-              << "; number of running containers: " << running_container_number;
-    return false;
+    for (std::size_t i = 0; i < services.size(); i++) {
```

Comment:
I think you already mentioned this, but I see two phases of optimization.

1) change only this method with something like
~~~
docker ps -q --filter=status=running --filter=label=com.docker.compose.project=shellhttpd --format='{{.Label "com.docker.compose.service" }} {{.Label "io.compose-spec.config-hash"}}'
httpd df03aff6219ace2d86e9003ac8cd294f3438d271c1614820ebe62432b07920b8
~~~

2) Call docker-ps one level up in the stack *once* with:
~~~
docker ps -q --filter=status=running --format='{{.Label "com.docker.compose.project"}} {{.Label "com.docker.compose.service" }} {{.Label "io.compose-spec.config-hash"}}'
fiotest fiotest 2fc5236772a8cc436c105493c5b8739d969e1aca3d90492382ea74719e6f18b3
shellhttpd httpd df03aff6219ace2d86e9003ac8cd294f3438d271c1614820ebe62432b07920b8
~~~

Point 2 will probably be a noticeable performance improvement.

Reply:
@doanac We've already discussed it with @ldts. The ultimate goal is what you outlined here https://github.com/foundriesio/aktualizr-lite/pull/46. In a nutshell, a new component `namespace Docker::Engine/Daemon` that exposes API to get required details of running containers as well as installed images (if needed) and using libcurl to make requests to the docker daemon. Then, ComposeAppEngine is initialized with an instance of Docker::Daemon or its factory method and makes request(s) to it just once before looping over apps in isRunning() method.

Reply:
right, if we are going to get rid of docker ps then it seemed simpler to me just parse this way and do a final parser in some other way when we are down to that... but I can optimize it here too

Reply:
anyway, fixed it too

Reply:
I get the impression, this might not be time for me to review this and i should wait a bit?

I assume this PR isn't going to move us to talking directly with the API though? i.e. - so not running `docker ps` in  loop is still worth doing since this will get into master?

Reply:
>I assume this PR isn't going to move us to talking directly with the API though? 
No, it's not.

But Jorge does one `docker ps` per app now, so it's not worse than it was before.

Reply:
@doanac I think it's ready for review.

Reply:
@doanac removed the docker ps stuff....at some point we should review and merge before I continue adding stuff...

Reply:
yep - i think its time to make this good enough and merge as well.